rp235x-hal-examples/src/bin/gpio_irq_example.rs

```diff
@@ -142,21 +144,23 @@ fn main() -> ! {
     }
 }
 
-#[interrupt]
+#[no_mangle]
+#[allow(non_snake_case)]
 fn IO_IRQ_BANK0() {
     // The `#[interrupt]` attribute covertly converts this to `&'static mut Option<LedAndButton>`
     static mut LED_AND_BUTTON: Option<LedAndButton> = None;
+    let led_and_button = unsafe { &mut *core::ptr::addr_of_mut!(LED_AND_BUTTON) };
```

Comment:
This `&mut *core::ptr::addr_of_mut!( ... )` IMHO doesn't make sense.

The whole raison d'être of `addr_of_mut!` is avoiding the `&mut`, because having a `&mut` (even for a short time, and not derefencing it) can potentially cause UB if you are not careful.

But here, you convert it to a `&mut` anyway. So I don't see any advantage of this difficult to read incantation, you can as well just do `unsafe { &mut LED_AND_BUTTON };` instead.

Now, that causes a warning. And this warning looks like it led you to writing the code above:

```
warning: creating a mutable reference to mutable static is discouraged
   --> src/bin/gpio_irq_example.rs:152:35
    |
152 |     let led_and_button = unsafe { &mut LED_AND_BUTTON };
    |                                   ^^^^^^^^^^^^^^^^^^^ mutable reference to mutable static
    |
    = note: for more information, see issue #114447 <https://github.com/rust-lang/rust/issues/114447>
    = note: this will be a hard error in the 2024 edition
    = note: this mutable reference has lifetime `'static`, but if the static gets accessed (read or written) by any other means, or any other reference is created, then any further use of this mutable reference is Undefined Behavior
    = note: `#[warn(static_mut_refs)]` on by default
help: use `addr_of_mut!` instead to create a raw pointer
    |
152 |     let led_and_button = unsafe { addr_of_mut!(LED_AND_BUTTON) };
    |                                   ~~~~~~~~~~~~~              +
```

I think the warning is valid, but the `help` is misleading. It even says "to create a raw pointer", and as written it would return a pointer, not a `&mut`, so the resulting code wouldn't build.

TL;DR:

```suggestion
    #[allow(static_mut_refs)]
    // Safety: As long as this interrupt is only enabled on a single core,
    // and the interrupt isn't somehow called reentrantly, there will not
    // be a second reference to LED_AND_BUTTON at the same time.
    let led_and_button = unsafe { &mut LED_AND_BUTTON };
```

Reply:
This will be a hard error in 2024 Edition, so I'll think of a better solution.